# Laboratory excercise 5

The notebook contains excercise connected to the auditory excerice 6. For any questions feel free to contact assistant: eda.jovicic@fer.hr

In [4]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# set seed for consistency
np.random.seed(2)

1. Load the cleaned dataset from the first exercise. If you haven't saved the dataset, rerun the exercise and save the final dataset.

In [5]:
#load data
df = pd.read_csv('cleaned_students_score.csv')
df

StudentId  Gender_female  Gender_male  LunchType_free/reduced  \
0         520645            1.0          0.0                     0.0   
1         303683            1.0          0.0                     0.0   
2         457351            1.0          0.0                     0.0   
3         812988            0.0          1.0                     1.0   
4         594678            0.0          1.0                     0.0   
...          ...            ...          ...                     ...   
29112     611241            1.0          0.0                     0.0   
29113     299810            0.0          1.0                     0.0   
29114     941589            1.0          0.0                     1.0   
29115     867225            1.0          0.0                     0.0   
29116     649362            0.0          1.0                     0.0   

       LunchType_standard  EthnicGroup_Label  ParentEduc_Label  \
0                     1.0                  2                 1   
1                     1.0                  2                 5   
2                     1.0                  1                 3   
3                     0.0                  0                 0   
4                     1.0                  2                 5   
...                   ...                ...               ...   
29112                 1.0                  3                 2   
29113                 1.0                  4                 2   
29114                 0.0                  2                 2   
29115                 1.0                  3                 0   
29116                 1.0                  1                 5   

       TestPrep_Label  ParentMaritalStatus_Label  PracticeSport_Label  \
0                   1                          1                    1   
1                   1                          1                    2   
2                   1                          2                    2   
3                   1                          1                    0   
4                   1                          1                    2   
...               ...                        ...                  ...   
29112               1                          2                    2   
29113               1                          2                    1   
29114               0                          1                    2   
29115               0                          1                    1   
29116               1                          1                    0   

       IsFirstChild_Label  TransportMeans_Label  WklyStudyHours_Label  \
0                       1                     1                     1   
1                       1                     2                     0   
2                       1                     1                     1   
3                       0                     2                     0   
4                       1                     1                     0   
...                   ...                   ...                   ...   
29112                   0                     1                     0   
29113                   0                     0                     0   
29114                   0                     0                     0   
29115                   0                     1                     0   
29116                   0                     1                     0   

       NrSiblings  MathScore  ReadingScore  WritingScore  
0             3.0         71            71            74  
1             0.0         69            90            88  
2             4.0         87            93            91  
3             1.0         45            56            42  
4             0.0         76            78            75  
...           ...        ...           ...           ...  
29112         2.0         59            61            65  
29113         1.0         58            53            51  
29114         1.0         61            70            67  
29115

2. Using linear regression, model the MathScore using features: TestPrep, NrSiblings, ParentMaritalStatus and ReadingScore. 

In [7]:
# Declares the model
mod = smf.ols(formula='MathScore ~ TestPrep_Label + NrSiblings + ParentMaritalStatus_Label + ReadingScore', data=df)
res = mod.fit()
res.summary()



<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.669
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                 1.474e+04
Date:                Thu, 04 Jan 2024   Prob (F-statistic):               0.00
Time:                        21:28:11   Log-Likelihood:            -1.0471e+05
No. Observations:               29117   AIC:                         2.094e+05
Df Residuals:                   29112   BIC:                         2.095e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=============================================================================================
                                coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------
Intercept                     6.3552      0.302     21.030      0.000       5.763       6.947
TestPrep_Label                1.1073      0.113      9.817      0.000       0.886       1.328
NrSiblings                    0.0154      0.036      0.426      0.670      -0.056       0.086
ParentMaritalStatus_Label    -0.0410      0.076     -0.536      0.592      -0.191       0.109
ReadingScore                  0.8570      0.004    239.501      0.000       0.850       0.864
==============================================================================
Omnibus:                      233.653   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              155.726
Skew:                          -0.010   Prob(JB):                     1.53e-34
Kurtosis:                       2.642   Cond. No.                         422.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Questions:

a) Using the summary, write down the formula with coefficients

$MathScore = $

𝑀𝑎𝑡ℎ𝑆𝑐𝑜𝑟𝑒 = 6.3552 + 1.1073 ∗ TestPrep + 0.0154 ∗ NrSiblings + (-0.0410) ∗ ParentMaritalStatus + 0.8570 ∗ ReadingScore



b) Are there any statisticaly important features?
TestPrep and ReadingScore 

c) Let's interpret the data, if students do test preparation, how does that affect the outcome on the math exam? 
If students did test preparation, their math exam score would be higher

d) Are students who practice sport better at math exams or worse? Using linear regression, model the MathScore using only one feature: PracticeSport. Don't include intercept. Using summary, decide which students do the best at math exams - those who never practice sport, sometimes practice sport or those who do it regularly?

In [9]:
# Declare the model
# - 1 je za uklanjanje intercepta. Intercept je inače uključen po defaultu a da se dodaje ručno izgledalo bi ovako + 1

mod = smf.ols(formula='MathScore ~ PracticeSport_Label - 1', data=df)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              MathScore   R-squared (uncentered):                   0.753
Model:                            OLS   Adj. R-squared (uncentered):              0.753
Method:                 Least Squares   F-statistic:                          8.857e+04
Date:                Thu, 04 Jan 2024   Prob (F-statistic):                        0.00
Time:                        21:39:14   Log-Likelihood:                     -1.4397e+05
No. Observations:               29117   AIC:                                  2.879e+05
Df Residuals:                   29116   BIC:                                  2.880e+05
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
PracticeSport_Label    38.3548      0.129    297.600      0.000      38.102      38.607
==============================================================================
Omnibus:                     1585.882   Durbin-Watson:                   1.658
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1060.414
Skew:                           0.352   Prob(JB):                    5.42e-231
Kurtosis:                       2.384   Cond. No.                         1.00
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

 Students who do best at math exams are those who practice sport regularly